File: content/07-trees/diameter_of_binary_tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def diameterOfBinaryTree(root: TreeNode) -> int:
    res = 0
    def dfs(node):
        nonlocal res
        if not node:
            return 0
        left = dfs(node.left)
        right = dfs(node.right)
        res = max(res, left + right)
        return 1 + max(left, right)
    dfs(root)
    return res

if __name__ == "__main__":
    r1 = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3))
    assert diameterOfBinaryTree(r1) == 3, "Test 1 Failed"
    r2 = TreeNode(1, TreeNode(2))
    assert diameterOfBinaryTree(r2) == 1, "Test 2 Failed"
    print("All tests passed!")

File: content/07-trees/diameter_of_binary_tree.js
class TreeNode {
    constructor(val = 0, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

function diameterOfBinaryTree(root) {
    let res = 0;
    function dfs(node) {
        if (!node) return 0;
        let left = dfs(node.left);
        let right = dfs(node.right);
        res = Math.max(res, left + right);
        return 1 + Math.max(left, right);
    }
    dfs(root);
    return res;
}

const assert = require('assert');
const r1 = new TreeNode(1, new TreeNode(2, new TreeNode(4), new TreeNode(5)), new TreeNode(3));
assert.strictEqual(diameterOfBinaryTree(r1), 3);
console.log("All tests passed!");
